File: LeetCode/HashMap/java/isomorphicStrings.java
package LeetCode.HashMap.java;

import java.util.HashMap;
import java.util.Map;

record isomorphicStringsRecord(String s, String t, boolean expected) {
}

public class isomorphicStrings {
    public boolean isIsomorphic(String s, String t) {
        Map<Character, Character> mapST = new HashMap<>();
        Map<Character, Character> mapTS = new HashMap<>();

        for (int i = 0; i < s.length(); i++) {
            Character c1 = s.charAt(i);
            Character c2 = t.charAt(i);

            if (mapST.containsKey(c1) && !mapST.get(c1).equals(c2) ||
                    mapTS.containsKey(c2) && !mapTS.get(c2).equals(c1)) {
                return false;
            }

            mapST.put(c1, c2);
            mapTS.put(c2, c1);
        }

        return true;
    }

    public static void main(String[] args) {
        isomorphicStringsRecord input = new isomorphicStringsRecord("egg", "add", true);
        isomorphicStrings.testSolution(input);

        input = new isomorphicStringsRecord("foo", "bar", false);
        isomorphicStrings.testSolution(input);

        input = new isomorphicStringsRecord("paper", "title", true);
        isomorphicStrings.testSolution(input);
    }

    private static void testSolution(isomorphicStringsRecord input) {
        System.out.println("Input: s: " + input.s());
        System.out.println("\tt: " + input.t());
        System.out.println("Expected: " + input.expected());
        boolean res = new isomorphicStrings().isIsomorphic(input.s(), input.t());
        System.out.println("Result: " + res);
        System.out.println(res == input.expected() ? "PASS" : "FAIL");
        System.out.println("-".repeat(50));
    }
}

File: LeetCode/HashMap/python/isomorphicStrings.py

class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        mapST = {}
        mapTS = {}

        for i in range(len(s)):
            c1 = s[i]
            c2 = t[i]

            if c1 in mapST and mapST.get(c1) != c2 or c2 in mapTS and mapTS.get(c2) != c1:
                return False
            
            mapST[c1] = c2
            mapTS[c2] = c1

        return True

    @staticmethod
    def testSolution(input) -> None:
        print("Input: s: " + input["s"])
        print("\tt: " + input["t"])
        print("Expected: " + str(input["expected"]))
        res = Solution().isIsomorphic(input["s"], input["t"])
        print("Result: " + str(res))
        print("PASS" if res == input["expected"] else "FAIL")
        print("-" * 50)

if __name__ == "__main__":
    input = {"s": "egg", "t": "add", "expected": True}
    Solution.testSolution(input)

    input = {"s": "foo", "t": "bar", "expected": False}
    Solution.testSolution(input)

    input = {"s": "paper", "t": "title", "expected": True}
    Solution.testSolution(input)